napari/layers/transforms.py

```diff
@@ -0,0 +1,173 @@
+from __future__ import annotations  # noqa: F407
+
+import toolz as tz
+from typing import Sequence
+import numpy as np
+
+# from ..utils.list import ListModel
+
+
+class Transform:
+    """Base transform class.
+
+    Defaults to the identity transform.
+
+    Parameters
+    ----------
+    func : callable, Coords -> Coords
+        A function converting an NxD array of coordinates to NxD'.
+    """
+
+    def __init__(self, func=tz.identity, inverse=None, name=None):
+        self.func = func
+        self._inverse_func = inverse
+        self.name = name
+
+        if func is tz.identity:
+            self._inverse_func = tz.identity
+
+    def __call__(self, coords):
+        """Transform input coordinates to output."""
+        return self.func(coords)
+
+    def set_slice(self, axes: Sequence[int]) -> Transform:
+        """Return a transform subset to the visible dimensions."""
+        raise NotImplementedError('Cannot subset arbitrary transforms.')
+
+    def set_pad(self, axes: Sequence[int]) -> Transform:
+        """Return a transform with added axes for non-visible dimensions."""
+        raise NotImplementedError('Cannot subset arbitrary transforms.')
+
+    @property
+    def inverse(self):
+        if self._inverse_func is not None:
+            return Transform(self._inverse_func, self.func)
+        else:
+            raise ValueError('Inverse function was not provided.')
+
+
+# class TransformChain(ListModel, Transform):
+#     def __init__(self, transforms=[]):
+#         super().__init__(
+#             basetype=Transform,
+#             iterable=transforms,
+#             lookup={str: lambda q, e: q == e.name},
+#         )
+#
+#     def __call__(self, coords):
+#         return tz.pipe(coords, *self)
+#
+#     def __newlike__(self, iterable):
+#         return ListModel(self._basetype, iterable, self._lookup)
+#
+#     def set_slice(self, axes: Sequence[int]) -> TransformChain:
+#         return TransformChain([tf.set_slice(axes) for tf in self])
+
+
+class STTransform(Transform):
+    """n-dimensional scale and translation (shift) class.
```

Comment:
I like a lot of the implementation details here, but overall I feel like this is a clear sign our model is being influenced far too much by vispy's own structure.

Reply:
That's fair, though I do think a `Scale` then `Translate` transform was a pretty sensible choice on their part for an abstraction. I'd like to add `AffineTransform` in a follow-up PR, at which point we could either leave this or replace it. For things like grid mode, you might only ever need `STTransform`, but we have to have a good way of chaining / composing. In the future we might want to add `PiecewiseAffine` or `Deformable` too, but I want to work towards those slowly